File: 1/2020.10.15/515.Find Largest Value in Each Tree Row/python/solution.py
class Solution:
    def largestValues(self, root: TreeNode) -> List[int]:
        # BFS
        if not root:
            return []
        res, q = [], [root]
        while q:
            size = len(q)
            level_max = float('-inf')
            for _ in range(size):
                node = q.pop(0)
                if node.val > level_max:
                    level_max = node.val
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            res.append(level_max)
        return res

    def largestValues(self, root: TreeNode) -> List[int]:
        # DFS
        def traversal(root, level):
            if not root:
                return
            if len(res) == level:
                res.append(float('-inf'))
            res[level] = max(res[level], root.val)
            traversal(root.left, level+1)
            traversal(root.right, level+1)
        res = []
        traversal(root, 0)
        return res

    # BFS from StefanPochmann
        maxes = []
        row = [root]
        while any(row):
            maxes.append(max(node.val for node in row))
            row = [kid for node in row for kid in (node.left, node.right) if kid]
        return maxes

File: 1/2020.10.15/515.Find Largest Value in Each Tree Row/java/solution.java
class Solution {

    List<Integer> res = new ArrayList<>();

    public List<Integer> largestValues(TreeNode root) {
        // DFS
        DFS(root, 0);
        return res;
    }

    private void DFS(TreeNode root, int level) {
        if (root == null) {
            return;
        }
        if (res.size() == level) {
            res.add(root.val);
        } else {
            res.set(level, Math.max(res.get(level), root.val));
        }
        DFS(root.left, level + 1);
        DFS(root.right, level + 1);
    }

    public List<Integer> largestValues(TreeNode root) {
        // BFS
        List<Integer> res = new ArrayList<>();
        if (root == null) {
            return res;
        }
        Deque<TreeNode> q = new LinkedList<>();
        q.offerLast(root);
        while (!q.isEmpty()) {
            int size = q.size();
            int max = 0;
            for (int i = 0; i < size; i++) {
                TreeNode node = q.pollFirst();
                max = i == 0 ? node.val : Math.max(max, node.val);
                if (node.left != null) {
                    q.offerLast(node.left);
                }
                if (node.right != null) {
                    q.offerLast(node.right);
                }
            }
            res.add(max);
        }
        return res;
    }
}
